### PR TITLE
Fix for AutoFriend for geyser extension

### PR DESCRIPTION
**Pull request by PizzaThatCodes**
Noticed that the Standalone has the option for AutoFriend and people wanted it on the extension one too, so here is a way for it, which also includes the api limitation for unfollow

### CODE REVIEW

**Comment by PizzaThatCodes** [comment]
Fixes Autofriend for geyser extension

File: bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java [hunk 1]
```diff
@@ -126,7 +126,37 @@ public void onPostInitialize(GeyserPostInitializeEvent event) {
 
             // Start the update timer
             GeyserImpl.getInstance().getScheduledThread().scheduleWithFixedDelay(this::tick, config.updateInterval, config.updateInterval, TimeUnit.SECONDS); // TODO Find API equivalent
+
+            // Due to API limitations stated in the config, need a separate update timer
+            GeyserImpl.getInstance().getScheduledThread().scheduleAtFixedRate(this::autoFriend, config.friendSyncConfig.updateInterval, config.friendSyncConfig.updateInterval, TimeUnit.SECONDS);
```
Comment (rtm516) [inline bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java:131]: This should only be added if one of the options is enabled. See https://github.com/rtm516/MCXboxBroadcast/blob/master/bootstrap/standalone/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/standalone/StandaloneMain.java#L101

File: bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java [hunk 2]
```diff
@@ -141,6 +171,7 @@ private void tick() {
             logger.error("Failed to update session information!", e);
         }
 
+
```
Comment (rtm516) [inline bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java:174]:
```suggestion
```

File: bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/ExtensionConfig.java [hunk 3]
```diff
@@ -9,9 +9,22 @@ public class ExtensionConfig {
...
     public boolean whitelistFriends;
+
+    public static class FriendSyncConfig {
+        @JsonProperty("update-interval")
+        public int updateInterval;
+        @JsonProperty("auto-follow")
+        public boolean autoFollow;
+        @JsonProperty("auto-unfollow")
+        public boolean autoUnfollow;
+    }
```
Comment (rtm516) [inline bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/ExtensionConfig.java:28]: This should be moved to a more generic shared class to save duplicate code

File: bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java [hunk 4]
```diff
@@ -126,7 +126,37 @@ public void onPostInitialize(GeyserPostInitializeEvent event) {
...
+                // Unfollow the person
+                if (config.friendSyncConfig.autoUnfollow && !person.isFollowingCaller && person.isFollowedByCaller) {
+                    logger.info("Removed " + person.displayName + " (" + person.xuid + ") as a friend");
+                    sessionManager.removeXboxFriend(person.xuid);
+                }
+            }
+        } catch (XboxFriendsException e) {
+            logger.error("Failed to sync friends", e);
+        }
+
```
Comment (rtm516) [inline bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java:159]: This should be moved to a more generic shared class to save duplicate code

**Review by rtm516 — CHANGES_REQUESTED** [review]

File: bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java [hunk 5]
```diff
@@ -68,7 +70,8 @@ public void onPostInitialize(GeyserPostInitializeEvent event) {
         }
 
         try {
-            config = new ObjectMapper(new YAMLFactory()).readValue(configFile, ExtensionConfig.class);
+            config = new ObjectMapper(new YAMLFactory()).configure(DeserializationFeature.FAIL_ON_IGNORED_PROPERTIES, false).readValue(configFile, ExtensionConfig.class);
+            friendConfig = new ObjectMapper(new YAMLFactory()).configure(DeserializationFeature.FAIL_ON_IGNORED_PROPERTIES, false).readValue(configFile, FriendConfig.class);
```
Comment (rtm516) [inline bootstrap/geyser/src/main/java/com/rtm516/mcxboxbroadcast/bootstrap/geyser/MCXboxBroadcastExtension.java:74]: The second `friendConfig` shouldn't be needed

File: core/src/main/java/com/rtm516/mcxboxbroadcast/core/FriendConfig.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package com.rtm516.mcxboxbroadcast.core;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.rtm516.mcxboxbroadcast.core.exceptions.XboxFriendsException;
+import com.rtm516.mcxboxbroadcast.core.models.FollowerResponse;
+
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class FriendConfig {
```
Comment (rtm516) [inline core/src/main/java/com/rtm516/mcxboxbroadcast/core/FriendConfig.java:9]: Remove annotation and maybe rename FriendUtils

File: core/src/main/java/com/rtm516/mcxboxbroadcast/core/FriendConfig.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+                    logger.info("Removed " + person.displayName + " (" + person.xuid + ") as a friend");
+                    sessionManager.removeXboxFriend(person.xuid);
+                }
+            }
+        } catch (XboxFriendsException e) {
+            logger.error("Failed to sync friends", e);
+        }
+
+    }
+
```
Comment (rtm516) [inline core/src/main/java/com/rtm516/mcxboxbroadcast/core/FriendConfig.java:48]:
```suggestion
    }
```

**Review by rtm516 — CHANGES_REQUESTED** [review]

**Comment by PizzaThatCodes** [comment]
When creating FriendUtils and removing the FriendConfig Variable from the one class it brought up an issue of trying to get the config's of the extension and standalone, so I moved their config files into a folder called "configs" inside the core directory which also includes the FriendSyncConfig to allow me to use those classes for the autoFriend() function.

**Comment by hasankayra04** [comment]
Any update on this?

**Comment by PizzaThatCodes** [comment]
Currently waiting on rtm516 to either push this to the main branch, or to comment any issues.


Currently if you want to use the jar for this version you can use:
https://ci.rtm516.co.uk/job/MCXboxBroadcast/view/change-requests/job/PR-21/

**Comment by onebeastchris** [comment]
Heyo! Unfortunately, ever since your last changes, i've been unable to actually get the friend listener to show up - Geyser logs the extension as existing & claims to laod it, however, it never prints anything in the console/starts up - this is now with the last build from the link; when you linked an initial build, that worked

**Comment by TBYT** [comment]
> update: works just fine; fault was on Geysers end - now fixed with [GeyserMC/Geyser@b48d4a0](https://github.com/GeyserMC/Geyser/commit/b48d4a0625ed9a269d81fea6e535345bce01c0af)

is it fixed with geyser by using this pr of mcxboxbroadcast or without this pr?

**Comment by onebeastchris** [comment]
The error wasn't coming from MCXboxBroadcast; it was on Geysers part and only in the edge case of running multiple extensions. This PR, and the normal version, should work just fine.

**Comment by TBYT** [comment]
ok nevermind then

**Comment by SonnenPingu** [comment]
Will it be possible to transfer this function to the 1.20 as well? It would be a shame if it was gone now.

**Comment by jefflima1** [comment]
Could you add this fix to new version? has been updated but the issue has not been merged

**Review by Adamiok — APPROVED** [review]
Looks fine to me.
Taking a long time to merge.

Sorry, wrong repo.

**Pull request merged**